addons/sale_stock_margin/tests/test_sale_stock_margin.py

```diff
@@ -200,3 +200,68 @@ def test_so_and_multicurrency(self):
         so.action_confirm()
         self.assertEqual(so_line.purchase_price, 200)
         self.assertEqual(so_line.price_unit, 400)
+
+    def test_so_and_multicompany(self):
+        """ In a multicompany environnement, when the user is on company C01 and confirms a SO that
+        belongs to a second company C02, this test ensures that the computations will be based on
+        C02's data"""
+        main_company = self.env['res.company']._get_main_company()
+        main_company_currency = main_company.currency_id
+        new_company_currency = self.env.ref('base.EUR') if main_company_currency == self.env.ref('base.USD') else self.env.ref('base.USD')
+
+        date = fields.Date.today()
+        self.env['res.currency.rate'].search([]).unlink()
+        self.env['res.currency.rate'].create([
+            {'currency_id': main_company_currency.id, 'rate': 1, 'name': date, 'company_id': False},
+            {'currency_id': new_company_currency.id, 'rate': 3, 'name': date, 'company_id': False},
+        ])
+
+        new_company = self.env['res.company'].create({
+            'name': 'Super Company',
+            'currency_id': new_company_currency.id,
+        })
+        self.env.user.company_id = new_company.id
+
+        self.pricelist.currency_id = new_company_currency.id
+
+        product = self._create_product()
+
+        incoming_picking_type = self.env['stock.picking.type'].search([('company_id', '=', new_company.id), ('code', '=', 'incoming')])
+        production_location = self.env['stock.location'].search([('company_id', '=', new_company.id), ('usage', '=', 'production')])
```

Comment:
I don't understand why you use the `production_location` as source location instead of `incoming_picking_type.default_location_src_id` here ?

Reply:
Because such a location does not exist
When creating the new company, several picking types are created. The Receipts picking type won't have any `default_location_src_id` defined: 
https://github.com/odoo/odoo/blob/63167ac8ae6ac08fa54a133ed9089dd872c7d73c/addons/stock/models/stock_warehouse.py#L932-L938
I wanted to use one of the available locations. From:
https://github.com/odoo/odoo/blob/24d52d4a8d0f475bb62b6653c08ce409e27e1d58/addons/stock/models/res_company.py#L176-L187
https://github.com/odoo/odoo/blob/24d52d4a8d0f475bb62b6653c08ce409e27e1d58/addons/stock/models/res_company.py#L158-L163
Here are the created locations (Name, Usage): 
`[('Super', 'view'), ('Stock', 'internal'), ('Inventory adjustment', 'inventory'), ('Production', 'production'), ('Scrap', 'inventory')]`
This is the reason why I chose the production one